1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,248 @@
+
+----
+Fred: Hey Sarah, there’s a new "fudge" field and it can be "blah" or "whatever"
+
+Sarah: Great! Thanks I’ll chuck that in now.
+----
+
+This works for a while. The API works, the app runs fine, we're making money, and we’re onto the next thing!
+
+This method has a huge number, the main being that communication is hard. You never really know if somebody understood what you meant even if they say they did. Beyond that, not having things written down means people can forget. If another client team also needs the information and they ask another client team, they are getting a copy of a copy which might be slightly wrong.
+
+Another outcome is the original author(s) cannot remember the exact rules, leading to time wasted checking the code. An even worse scenario is that the original author(s) are not reachable for some reason. Maybe the whole team of developers who worked on that API are all on a flight together from New York to Uruguay, and for that entire 9 hour flight nobody can get any answers about how the API is supposed to work, so they're stuck not being able to fix a production issue...
+
+In this instance two common things happen.
+
+1. Developers waste time writing a new endpoint out of fear of using the old one
+2. Developers waste time trying to guess the contract, and might guess wrong
+
+=== Shoddy Human Contracts
+
+When folks first start thinking about contracts, it often takes the form of rudimentary API Reference Documentation. This might be dumped into a Google Doc, shoved in a wiki, or written up in finger blistering HTML. These formats are usually fairly painstaking to create, because you are focusing on formatting _and_ content, shuffling things around whenever you remember that you should list headers too, and copying and pasting examples of JSON which might change over time. API Developers would not settle for writing software like this, but apparently when it comes to documentation it's accepted as the norm.
+
+This whole slow, manual approach is probably why many people forgo writing contracts this way, and opt for verbal contracts instead.It probably would not be a huge jump to say the "startup mindset" (and "agile") are one of a few potential driving factors in this "just get on with it" approach. There are many technical leaders inexperienced in this area telling their team "We'll write lovely docs later when we've got more time!" , without realizing their lack of clarity is slowing down initial development, client integrations, increasing testing efforts, and causing bizarre production issues and throughout their ecosystem.
+
+.It's pretty common to blame REST for not forcing a type system onto everyone involved in the API, but they're absolutely useable. Don't blame a paradigm for implementation level decisions.
```

Comment:
Don't think this comment / image belongs here? Sounds like something for the "paradigms" section?

Reply:
I wanted to work that meme in somewhere, I'll find somewhere else.